Add BoundingValue Damper and support for NodalDamper

bwspenc opened the pull request:
Closes #7856

bwspenc committed:
Add BoundingValueNodalDamper, BoundingValueElementDamper, and NodalDamper support ref #7856

bwspenc commented:
A few comments/questions about this:
- The code in NonlinearSystem and MooseVariable for computing the increment needs to different things for nodal and element dampers. What I've done doesn't feel like a very elegant solution.
- The solution increment vector seems to actually be the negative of the solution increment. I could set the value of _increment to the negative of that value when I'm setting it, or use the negative of _increment when I'm using it if the sign matters in the damper. I did the latter, but it could be done either way.
- Both the BoundingValue and the MaxIncrement Damper could be implemented as either nodal or element dampers. I'm not sure whether it makes any sense to make versions of these that do it both ways, e.g. ElementBoundingValue, NodalBoundingValue. My opinion is that's probably getting a little carried away.
- I added a bit of code that prints to the console the value of the damping is if it's ever less than 1. That seems pretty helpful for the user to understand what's going on.
- I still need to add a test for this, which is a little tricky, but I have ideas. I wanted to get feedback on the code first.

andrsd commented on framework/src/base/MooseVariable.C:
```diff
@@ -1605,6 +1605,18 @@ MooseVariable::computeIncrement(const NumericVector<Number> & increment_vec)
+  _increment.resize(1);
+
+  // Compute the increment for the current DOF
+  _increment[0] =  increment_vec(_dof_indices[0]);
```
one space after `=`

bwspenc replied:
Fixed

andrsd commented on framework/src/base/NonlinearSystem.C:
```diff
@@ -692,6 +697,8 @@ NonlinearSystem::addDamper(const std::string & damper_name, const std::string &
     else if (ed)
       _element_dampers.addObject(ed,tid);
+    else if (nd)
+      _nodal_dampers.addObject(nd,tid);
```
This is a copy and paste, but add a space after `,`.

bwspenc replied:
Fixed

andrsd commented on framework/src/base/ComputeNodalDampingThread.C:
```diff
@@ -0,0 +1,79 @@
+  for (const auto & damper : ndampers)
+    damped_vars.insert(damper->getVariable());
+
+  _nl.reinitIncrementAtNodeForDampers(_tid,damped_vars);
```
Add a space after  `,`.

bwspenc replied:
Fixed

andrsd commented on framework/src/base/ComputeElemDampingThread.C:
```diff
@@ -47,7 +47,14 @@ ComputeElemDampingThread::onElement(const Elem *elem)
+  for (const auto & damper : edampers)
+    damped_vars.insert(damper->getVariable());
+
+  _nl.reinitIncrementAtQpsForDampers(_tid,damped_vars);
```
A space after `,`

bwspenc committed:
Add test of BoundingValueNodalDamper and BoundingValueElementDamper closes #7856

bwspenc commented:
I addressed the comments of @andrsd and added a test.

andrsd commented on test/tests/dampers/bounding_value/bounding_value_test.i:
```diff
@@ -0,0 +1,85 @@
+#  [./TimeStepper]
+#    type = ConstantDT
+#    min_dt = 0.5
+#  [../]
```
Let's remove this commented out section

andrsd commented on test/tests/dampers/bounding_value/bounding_value_test.i:
```diff
@@ -0,0 +1,85 @@
+[]
+
+[Variables]
+  active = 'u'
```
I'd remove this

andrsd commented on test/tests/dampers/bounding_value/bounding_value_test.i:
```diff
@@ -0,0 +1,85 @@
+[]
+
+[BCs]
+  active = 'left right'
```
Remove this as well

andrsd commented:
After cleaning up the input file, this looks good to me.

@friedmud Since you built dampers, I would not mind a second opinion on this.  What @bwspenc did is he mimicked what we do for elemental dampers, but he does it for nodal values.

bwspenc committed:
Fix a couple of minor issues in MaxIncrement damper test ref #7856

When using this test as a template for a test of a new damper,
I noticed that it was incorrectly using a first order variable
on a higher order mesh. Also make other minor changes.

bwspenc commented:
@andrsd I made the changes you suggested to that test. While I was at it, I made the same changes to the test that I based it on, updated the comments at the top of the input file, and made a second version of the test that tests the min_value (the original one just tested max_value).

andrsd reviewed (approved):
I will let this sit here till like lunch. If nobody objects, I will merge then...

dschwen commented:
Shouldn't this be called `BoundingValueNodalDamper` rather than `BoundingValue`? The latter is awfully general. We have an element damper in Marmot that does almost the same thing (but more specific) and could be replaced by a more general `BoundingValueElementDamper`.

bwspenc commented:
@dschwen I was thinking the same thing (although I wouldn't put `Damper` at the end: how about `BoundingNodalValue`?).  Along those lines, it seems like the existing `MaxIncrement` Damper shoulde also be re-named (which will be a pain because it's so widely used), and that there probably should be a nodal version of that. That's an issue for another time, but I'm happy to re-name this.

bwspenc commented:
@dschwen per our discussion, I renamed this to `BoundingValueNodalDamper` and added `BoundingValueElementDamper`, as well as versions of the tests for the new element version of this.
